otlp/common.go

```diff
@@ -402,33 +402,56 @@ func getMarshallableValue(value *common.AnyValue) interface{} {
 	return nil
 }
 
-// This function returns a value that can be handled by Honeycomb -- it must be one of:
-// string, int, bool, float. All other values are converted to strings containing JSON.
-func getValue(value *common.AnyValue) (result interface{}, truncatedBytes int) {
+func flatten(result map[string]interface{}, key string, value *common.AnyValue, depth int) (truncatedBytes int) {
 	switch value.Value.(type) {
 	case *common.AnyValue_StringValue:
-		return value.GetStringValue(), 0
+		result[key] = value.GetStringValue()
 	case *common.AnyValue_BoolValue:
-		return value.GetBoolValue(), 0
+		result[key] = value.GetBoolValue()
 	case *common.AnyValue_DoubleValue:
-		return value.GetDoubleValue(), 0
+		result[key] = value.GetDoubleValue()
 	case *common.AnyValue_IntValue:
-		return value.GetIntValue(), 0
-	// These types are all be marshalled to a string after conversion to Honeycomb-safe values.
-	// We use our limitedWriter to ensure that the string can't be bigger than the allowable,
-	// and it also minimizes allocations.
-	// Note that an Encoder emits JSON with a trailing newline because it's intended for use
-	// in streaming. This is correct but sometimes surprising and the tests need to expect it.
-	case *common.AnyValue_ArrayValue, *common.AnyValue_KvlistValue, *common.AnyValue_BytesValue:
-		arr := getMarshallableValue(value)
-		w := newLimitedWriter(fieldSizeMax)
-		enc := json.NewEncoder(w)
-		err := enc.Encode(arr)
-		if err == nil {
-			return w.String(), w.truncatedBytes
+		result[key] = value.GetIntValue()
+	case *common.AnyValue_BytesValue:
+		truncated := addToMapAsJson(result, key, value)
+		truncatedBytes += truncated
+	case *common.AnyValue_ArrayValue:
+		for i, entry := range value.GetArrayValue().Values {
```

Comment:
Does it give us any qualms that this means that it becomes impossible to search an array for a value? I don't honestly know which is worse. Another option here is that we could add a DC function to let someone extract an indexed element of a JSON array.

Reply:
Another downside here is that we could potentially pollute a dataset with a thousand new columns when they never had any need for that data in the first place. KVLists are named, so that's at least more expected.

Hmmm. I'm more concerned with that downside than being able to search an array (as a string) for a value. Will have to let this simmer a little bit. We can probably proceed right now and then make a call after evaluating all the tradeoffs?

Reply:
Decided to only do kv lists for now. We can revisit arrays later in #239.